lib/runtime/src/ops/concat.cc

```diff
@@ -16,537 +16,199 @@
-  PROFILING
-}
+using namespace FlexFlow::Kernels::Concat;
 
-// declare Legion names
-using Legion::ArgumentMap;
 using Legion::Context;
-using Legion::coord_t;
-using Legion::Domain;
-using Legion::FutureMap;
-using Legion::IndexLauncher;
 using Legion::PhysicalRegion;
-using Legion::Predicate;
-using Legion::Rect;
-using Legion::RegionRequirement;
 using Legion::Runtime;
 using Legion::Task;
-using Legion::TaskArgument;
-using Legion::TaskLauncher;
-using PCG::Node;
 
-using namespace FlexFlow::Kernels::Concat;
+enum Slots {
+  INPUTS,
+  OUTPUT,
+  ATTRS,
+  PROFILING,
+  HANDLE,
+  PER_DEVICE_STATE,
+  NUM_INPUTS
+};
 
-bool operator==(ConcatParams const &lhs, ConcatParams const &rhs) {
-  return lhs.axis == rhs.axis;
-}
+OpTaskInvocation init(ConcatAttrs const &attrs) {
+  OpTaskBinding binding;
 
-ConcatParams Concat::get_params() const {
-  ConcatParams params;
-  params.axis = legion_axis;
-  return params;
-}
+  binding.bind_arg(PROFILING, profiling_settings());
+  binding.bind_arg(ATTRS, attrs);
 
-Tensor
-    FFModel::concat(int n, Tensor const *tensors, int axis, char const *name) {
-  Layer *concat = new Layer(this,
-                            OP_CONCAT,
-                            DT_FLOAT,
-                            name,
-                            n /*inputs*/,
-                            0 /*weights*/,
-                            1 /*outputs*/,
-                            tensors);
-  int numdim = tensors[0]->num_dims;
-  // Making sure axis is between [0, numdim)
-  axis = (axis % numdim + numdim) % numdim;
-  int dims[MAX_TENSOR_DIM];
-  for (int i = 0; i < numdim; i++) {
-    dims[i] = tensors[0]->dims[i];
-  }
-  for (int i = 1; i < n; i++) {
-    assert(tensors[i]->data_type == tensors[0]->data_type);
-    assert(tensors[i]->num_dims == tensors[0]->num_dims);
-    for (int j = 0; j < numdim; j++) {
-      if (j != numdim - axis - 1) {
-        assert(tensors[i]->dims[j] == tensors[0]->dims[j]);
-      } else {
-        dims[j] += tensors[i]->dims[j];
-      }
-    }
-  }
-  concat->outputs[0] = create_tensor_legion_ordering(
-      numdim, dims, tensors[0]->data_type, concat, 0, true /*create_grad*/);
-  concat->add_int_property("legion_axis", numdim - axis - 1);
-  layers.push_back(concat);
-  return concat->outputs[0];
+  return {CONCAT_INIT_TASK_ID, binding};
 }
 
-Op *Concat::create_operator_from_layer(
-    FFModel &model,
-    Layer const *layer,
-    std::vector<ParallelTensor> const &inputs) {
-  long long value;
-  layer->get_int_property("legion_axis", value);
-  int legion_axis = value;
-  return new Concat(
-      model, inputs.size(), inputs.data(), legion_axis, layer->name);
-}
+OpTaskInvocation forward(ConcatAttrs const &attrs) {
+  OpTaskBinding binding;
+  binding.bind_arg(PER_DEVICE_STATE,
+                   per_device_op_state<ConcatPerDeviceState>());
+  binding.bind(INPUTS, get_input_tensors());
+  binding.bind(OUTPUT, output_tensor(0));
+  binding.bind(NUM_INPUTS, get_number_inputs());
```

Comment:
Same here. `get_number_inputs()` doesn't give us any information. If anything, the number of inputs should be a field in `ConcatAttrs`